ts/session/apis/open_group_api/sogsv3/sogsBlinding.ts

```diff
@@ -0,0 +1,428 @@
+  return data;
+};
+
+async function decryptWithSessionBlindingProtocol(
+  data: Uint8Array,
+  isOutgoing: boolean,
+  otherBlindedPublicKey: string,
+  serverPubkey: string,
+  userEd25519KeyPair: ByteKeyPair
+) {
+  const sodium = await getSodiumRenderer();
+  if (data.length <= sodium.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES) {
+    throw new Error(
+      `data is too short. should be at least ${sodium.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES} but is ${data.length}`
+    );
+  }
+
+  const blindedKeyPair = SogsBlinding.getBlindingValues(
+    from_hex(serverPubkey),
+    userEd25519KeyPair,
+    sodium
+  );
+  if (!blindedKeyPair) {
+    throw new Error('Decryption failed');
+  }
+  /// Step one: calculate the shared encryption key, receiving from A to B
+  const otherKeyBytes = from_hex(PubKey.removePrefixIfNeeded(otherBlindedPublicKey));
+  const kA = isOutgoing ? blindedKeyPair.publicKey : otherKeyBytes;
+  const decKey = sharedBlindedEncryptionKey({
+    secretKey: userEd25519KeyPair.privKeyBytes,
+    otherBlindedPublicKey: otherKeyBytes,
+    fromBlindedPublicKey: kA,
+    toBlindedPublicKey: isOutgoing ? otherKeyBytes : blindedKeyPair.publicKey,
+    sodium,
+  });
+  if (!decKey) {
+    throw new Error('Decryption failed');
+  }
+
+  // v, ct, nc = data[0], data[1:-24], data[-24:]
+  const version = data[0];
+  const ciphertext = data.slice(
+    1,
+    data.length - sodium.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES
+  );
+  const nonce = data.slice(data.length - sodium.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES);
+
+  // Make sure our encryption version is okay
+
+  if (version !== 0) {
+    throw new Error('Decryption failed');
+  }
+
+  // Decrypt
+  const innerBytes = sodium.crypto_aead_xchacha20poly1305_ietf_decrypt(
+    null,
+    ciphertext,
+    null,
+    nonce,
+    decKey
+  );
+  if (!innerBytes) {
+    throw new Error('Decryption failed');
+  }
+  const numBytesPubkey = PubKey.PUBKEY_LEN_NO_PREFIX / 2;
+  // Ensure the length is correct
+  if (innerBytes.length <= numBytesPubkey) {
+    throw new Error('Decryption failed');
+  }
+
+  // Split up: the last 32 bytes are the sender's *unblinded* ed25519 key
+  const plainText = innerBytes.slice(0, innerBytes.length - numBytesPubkey);
+  const senderEdpk = innerBytes.slice(innerBytes.length - numBytesPubkey);
+
+  // Verify that the inner sender_edpk (A) yields the same outer kA we got with the message
+  const blindingFactor = generateBlindingFactor(serverPubkey, sodium);
+  const sharedSecret = combineKeys(blindingFactor, senderEdpk, sodium);
+
+  if (!isEqual(kA, sharedSecret)) {
+    throw new Error('Invalid Signature');
+  }
+  // Get the sender's X25519 public key
+  const senderSessionIdBytes = toX25519(senderEdpk, sodium);
+
+  return { plainText, senderUnblinded: `${KeyPrefixType.standard}${to_hex(senderSessionIdBytes)}` };
+}
+
+/**
+ * This function is actually just used for testing and is useless IRL.
```

Comment:
Shouldn't we move this into the unit test where it's used?

Reply:
moved to tests